### PR TITLE
Add CRAM reader (alpha)

### PR DESCRIPTION
**Pull request by athos**
This PR implements an experimental CRAM reader.

At this moment, the API only provides the following basic features:
- create a new CRAM reader for the specified file
- create a cloned CRAM reader from another CRAM reader
- reads all the alignment as a lazy sequence

The namespace structure looks like:

| Namespace | Description |
| :--- | :--- |
| `cljam.io.cram` | Provides the pubilc API |
| `cljam.io.cram.decode.data-series` | Implements decoders for data series and tags |
| `cljam.io.cram.decode.record` | Implements decoder for CRAM records |
| `cljam.io.cram.decode.structure` | Implements decoders for decoding CRAM constructs such as containers, slices, blocks, etc. |
| `cljam.io.cram.core` | Provides the CRAM reader constructors |
| `cljam.io.cram.reader` | Implements the CRAM reader |
| `cljam.io.cram.seq-resolver` | Provides an implementation of the "seq resolver", which abstracts out the direct dependency on the sequence reader from within the implementation of CRAM reader (to resolve some cyclic dependency) |
| `cljam.io.cram.seq-resolver.protocol` | provides the protocol definition of the "seq resolver" |

Here are the steps I suggest for reading the code:
- `cljam.io.cram.decode.structure(-test)`: grasp how the CRAM constructs are decoded and what kind of information they contain
- `cljam.io.cram.decode.data-series(-test)`: grasp how the data series decoders work
- `cljam.io.cram.decode.record(-test)`: get the big picture of how CRAM record decoder coordinates with the data series decoders
- others

## Limitations

This implementation is quite naive, aiming to make it easier to reason about its compliance with the CRAM specification, and it's not primarily focused on efficiency. Additionally, it currently supports only a very limited subset of the specification's features. Here are the current limitations (which might not be exhaustive):

- No support for region access or index access
- Decoding always requires a reference file
  - According to the CRAM specification, it should be possible to read a CRAM file without a reference file if the reference sequences are embedded in the CRAM file or if the alignments are unmapped
- The decoder never generates QNAMEs even if they are not embedded in the CRAM file
  - According to the CRAM specification, it's the decoder's responsibility to generate QNAMEs if they are not embedded in the file
- Does not support decoding from core data blocks
- Currently does not support the following compression methods and encodings:
  - Compression methods:
    - rans4x16
    - adaptive arithmetic coder
    - fqzcomp
    - name tokenizer
  - Encodings:
    - Huffman (only supported when it's a 1-alphabet/0-bit encoding)
    - Beta
    - Subexponential
    - Gamma
    - Golomb (deprecated)
    - Golomb-Rice (deprecated)
- The `RG` tag will always be added to the decoding result if the encoded record has the read group info
- Does not automatically calculate the `MD`/`NM` tags from read features
- Does not perform any CRC/MD5 checks
- Does not support coordination among more than two mate reads
- Does not support reading containers with a header larger than 256 bytes

Despite these limitations, it's still possible to read actual CRAM files that are used in practice.

### CODE REVIEW

File: src/cljam/io/cram.clj [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+(defn reader
+  "Creates a CRAM reader depending on the argument f: If f is a file or a string
+  that representing the path to a CRAM file, returns a new reader that reads
+  that CRAM file. If f is a CRAM reader, creates and returns a cloned CRAM reader
+  from it.
+
+  The function also takes an optional argument `option`, which is a map that
+  consists of:
+    - reference: a string representing the path to a reference file"
```
Comment (alumi) [inline src/cljam/io/cram.clj:16]: I think it would be beneficial to document that an instance of `ISequenceReader` can also be passed, not just a string.

**Review by alumi — COMMENTED** [review]
Thank you for implementing the CRAM decoder!
I understand that interpreting the CRAM spec might have been challenging due to their complexity. There were no issues found with the decoder's behavior.
I've added a few comments, so please review them when you have a moment.
Thank you for your hard work! 👍 👍

File: test/cljam/io/cram/decode/structure_test.clj [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+              {:method 1, :content-type 4, :content-id 28, :size 23, :raw-size 3, :crc [186 219 248 202]}
+              {:method 1, :content-type 4, :content-id 29, :size 20, :raw-size 0, :crc [111 73 18 0]}
+              {:method 1, :content-type 4, :content-id 30, :size 20, :raw-size 0, :crc [82 112 247 118]}
+              {:method 4, :content-type 4, :content-id 5063770, :size 10053, :raw-size 43243, :crc [246 219 153 220]}
+              {:method 4, :content-type 4, :content-id 5131619, :size 1492, :raw-size 7923, :crc [161 9 102 22]}]
+             (mapv (fn [_]
+                     (-> (struct/decode-block bb)
+                         (update :crc bytes->vec)
+                         (dissoc :data)))
+                   (range 29))))
```
[the remark this answers is not in the document]
Comment (athos) [reply test/cljam/io/cram/decode/structure_test.clj:150]: Thank you for checking the test data in detail. At this moment, I can't come up with the cause of the difference, so I'll take a look into it later.

File: src/cljam/io/cram.clj (hunk 1 above)
Comment (athos) [reply src/cljam/io/cram.clj:16]: Will do 💪

**Review by niyarin — APPROVED** [review]
Thank you for adding the CRAM decode feature.
LGTM.

File: test/cljam/io/cram/decode/structure_test.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (athos) [reply test/cljam/io/cram/decode/structure_test.clj:150]: Sorry for the late reply! After investigating this issue, it seems likely that this is a bug related to htsjdk's CRAM encoder.

The number of blocks stored in both the container header and the slice header appears to be interpreted as the number of blocks contained within that particular container and slice, respectively. Notably, slices themselves do NOT include compression header blocks or slice header blocks.

htslib counts the compression header block and slice header block in the total block count for the container header ([1], [2]) while it only includes the number of data blocks in the slice header's block count ([3]). This interpretation aligns with [the test data provided by hts-specs](https://github.com/samtools/hts-specs/tree/26347448cadff3cf40982d60fe2a97f20d2543ea/test/cram/3.0/passed). So, a rational interpretation here would be that the block count is 31 for the container header and 29 for the slice header.

In htsjdk, on the other hand, both the container header's and the slice header's block counts are calculated as the number of data blocks including core data blocks ([4]) resulting in a block count of 29 for both.

27d1c69 has corrected the block counts in the container header of the CRAM files generated using htsjdk according to the above interpretation.

[1]: https://github.com/samtools/htslib/blob/78e507dbd8a0567c7f3c8c1e265d36218e3f0e77/cram/cram_encode.c#L2154
[2]: https://github.com/samtools/htslib/blob/78e507dbd8a0567c7f3c8c1e265d36218e3f0e77/cram/cram_encode.c#L2477
[3]: https://github.com/samtools/htslib/blob/78e507dbd8a0567c7f3c8c1e265d36218e3f0e77/cram/cram_encode.c#L1233-L1234
[4]: https://github.com/samtools/htsjdk/blob/b16416c078e2df3e3a21486207d0c0c31b293cb9/src/main/java/htsjdk/samtools/cram/structure/Container.java#L93-L104

[the remark this answers is not in the document]
Comment (athos) [reply test/cljam/io/cram/decode/structure_test.clj:150]: Sorry, I forgot to update `test.cram`, not only `medium*.cram`!  5d3d289 has superseded 27d1c69 now.

**Review by alumi — APPROVED** [review]
Thank you for the investigation and the update!
I confirmed that the number of blocks in the CRAM files are reasonable now.
LGTM 👍 👍

**Pull request merged**